### PR TITLE
Support processing HTTP/1.1 headers as they arrive

### PR DESCRIPTION
**Pull request by gdwoolbert3**
## Introduction

This is my attempt at implementing the feature requested in [Issue 452](https://github.com/elixir-mint/mint/issues/452).

The motivation for the change was the `X-ClickHouse-Progress` header system used by the [ClickHouse HTTP interface](https://clickhouse.com/docs/concepts/features/interfaces/http) (which I strongly suspect was the reason behind the aforementioned issue).

## Implementation

This PR adds a new `:stream_headers` option for HTTP/1.1 connections that allows headers and trailers to be emitted as responses as soon as they're available instead of buffering them until they've all arrived. A few nuances about enabling this feature:
- All _complete_ headers from a single chunk are emitted at the same time
- It's theoretically possible to receive **NO** `:headers` responses for a given request whereas, when not streaming headers, an empty list would be emitted. This change is explicitly laid out in the `Mint.HTTP` documentation.

This feature is opt-in, meaning there should be no changes for existing users.

## Tesing

All existing tests pass and new stream headers tests were added to the HTTP/1.1 connection test module.

Open to any and all feedback 😃

### CODE REVIEW

**Comment by ericmj** [comment]
The option and response shape make sense, but there’s one parsing case to fix before merging. If a socket message ends exactly after a complete header line, such as `X-ClickHouse-Progress: ...\r\n`, `HTTP1.stream/2` buffers the header and returns no response until another byte arrives. This leaves progress updates one socket message behind, and a lone update won’t be emitted until the next header or the end of the section.

Could you add tests where a regular header and a trailer end at that boundary, with no byte from the following line, and update the parser to emit them immediately?

**Comment by gdwoolbert3** [comment]
@ericmj nice catch! If I'm understanding correctly, this appears to be a quirk of the underlying `:erlang.decode_packet/3` function:

```
iex> :erlang.decode_packet(:httph_bin, "X-ClickHouse-Progress: foo\r\n", [])
{:more, :undefined}

iex> :erlang.decode_packet(:httph_bin, "X-ClickHouse-Progress: foo\r\nX", [])
{:ok, {:http_header, 0, "X-Clickhouse-Progress", "X-ClickHouse-Progress", "foo"}, "X"}
```

With that being said, the ideal fix isn't particularly clear to me. If I'm reading RFC 7230 right, I _think_ we could naively check that the incoming data ends with `\r\n` and then do something to force the header to parse (ie appending a dummy byte), but that doesn't feel like the cleanest solution.

Do you have any suggestions? If not, I'll go with the aforementioned approach.

On a related note, do you have insight into whether or not the above behavior is intentional?

**Comment by wojtekmach** [comment]
one reason might be HTTP1 line folding, a header could have spanned multiple lines, https://www.rfc-editor.org/info/rfc9112/#section-5.2,

```elixir
iex> :erlang.decode_packet(:httph_bin, "header: chunk1\r\n chunk2\r\n\r\n", [])
{:ok, {:http_header, 0, "Header", "header", "chunk1\r\n chunk2"}, "\r\n"}
```

and so when getting `"header: chunk1\r\n"`, returning :more makes sense. That particular behaviour is obsolete in the spec however. And so yeah, it would be nice if we've gotten `{:ok, {:http_header, ...}, ""}` instead, but that'd probably require adding an option to packet parser and that probably has very high barrier to entry.

**Comment by ericmj** [comment]
Let's retry the `:erlang.decode_packet` call with `data <> sentinel` when it returns `:more`, `stream_headers` is enabled, and `data` ends in LF. Where `sentinel` is a single byte that doesn't trigger the legacy line folding.

**Comment by gdwoolbert3** [comment]
[Pushed](ff47c8315322b9b2556e8a02e8aab3c9589d1b22) the fix recommended by @ericmj.

As it turns out, `:erlang.decode_packet/3` allows LF in addition to CRLF as a header sep:

```
iex> :erlang.decode_packet(:httph_bin, "X-ClickHouse-Progress: foo\r\n", [])
{:more, :undefined}
iex> :erlang.decode_packet(:httph_bin, <<"X-ClickHouse-Progress: foo\r\n", 0>>, [])
{:ok, {:http_header, 0, "X-Clickhouse-Progress", "X-ClickHouse-Progress", "foo"}, <<0>>}
iex> :erlang.decode_packet(:httph_bin, <<"X-ClickHouse-Progress: foo\n", 0>>, [])
{:ok, {:http_header, 0, "X-Clickhouse-Progress", "X-ClickHouse-Progress", "foo"}, <<0>>}
```

I kept this behavior in order to maintain parity with the underlying functionality, but I'm happy to update at your discretion.

File: lib/mint/http1.ex [hunk 1]
```diff
@@ -1004,6 +1004,25 @@ defmodule Mint.HTTP1 do
...
+
+  defp do_decode_header(data, true) do
+    # By default, :erlang.decode_packet/3 asks for more data when a packet
+    # containing a full header ends with a line feed (likely to handle line
+    # folding). If we get a :more response on a packet that ends with a line
+    # feed, we append a sentinel byte and attempt to decode again.
+    with :more <- Response.decode_header(data) do
+      data_size = byte_size(data)
+
+      with <<_::binary-size(data_size - 1), 10>> <- data,
```
Comment (ericmj) [inline lib/mint/http1.ex]:
```suggestion
      with <<_::binary-size(data_size - 1), ?\n>> <- data,
```

File: lib/mint/http1.ex [hunk 2]
```diff
@@ -963,6 +1004,25 @@ defmodule Mint.HTTP1 do
...
+           {:ok, {name, value}, rest} <- Response.decode_header(<<data::binary, 0>>) do
+        {:ok, {name, value}, binary_part(rest, 0, byte_size(rest) - 1)}
+      else
+        _ -> :more
```
Comment (ericmj) [inline lib/mint/http1.ex]: `MIX_ENV=test mix compile --force --warnings-as-errors` fails because `data_size` is not pinned in the bitstring size expression. This version also preserves the retry error and removes the added byte only when the remainder matches the sentinel.

```suggestion
      with <<_::binary-size(^data_size - 1), ?\n>> <- data do
        case Response.decode_header(<<data::binary, 0>>) do
          {:ok, {name, value}, <<0>>} -> {:ok, {name, value}, ""}
          result -> result
        end
      else
        _ -> :more
```

File: lib/mint/http1.ex [hunk 3]
```diff
@@ -963,6 +1004,25 @@ defmodule Mint.HTTP1 do
     end
   end
 
+  defp do_decode_header(data, false), do: Response.decode_header(data)
+
+  defp do_decode_header(data, true) do
```
Comment (ericmj) [inline lib/mint/http1.ex]:
```suggestion
  defp do_decode_header(data, true = _stream_headers) do
```

File: test/mint/http1/conn_test.exs [hunk 4]
```diff
@@ -1273,6 +1273,167 @@ defmodule Mint.HTTP1Test do
...
+      {:ok, conn, ref} = HTTP1.request(conn, "GET", "/", [], nil)
+
+      assert {:ok, conn, [_status]} =
+               HTTP1.stream(conn, {:tcp, conn.socket, "HTTP/1.1 200 OK\r\n"})
+
+      # Send a complete header ending exactly at packet boundary (no subsequent bytes)
+      assert {:ok, conn, responses} =
+               HTTP1.stream(conn, {:tcp, conn.socket, "X-Progress: 50\r\n"})
+
+      assert [{:headers, ^ref, [{"x-progress", "50"}]}] = responses
```
Comment (ericmj) [inline test/mint/http1/conn_test.exs]: Use two headers here while keeping the final LF at the packet boundary. This covers the recursive path where the first header has real lookahead and only the final header uses the sentinel.

```suggestion
      assert {:ok, conn, responses} =
               HTTP1.stream(
                 conn,
                 {:tcp, conn.socket, "X-Progress: 50\r\nX-Other: value\r\n"}
               )

      assert [{:headers, ^ref, [{"x-progress", "50"}, {"x-other", "value"}]}] = responses
```

File: lib/mint/http1.ex [hunk 5]
```diff
@@ -1004,6 +1004,25 @@ defmodule Mint.HTTP1 do
     end
   end
 
+  defp do_decode_header(data, false), do: Response.decode_header(data)
```
Comment (ericmj) [inline lib/mint/http1.ex]:
```suggestion
  defp do_decode_header(data, false = _stream_headers), do: Response.decode_header(data)
```

**Review by ericmj — CHANGES_REQUESTED** [review]

File: lib/mint/http1.ex (hunk 3 above)
Comment (gdwoolbert3) [reply lib/mint/http1.ex]: 5940b9a68d479de238e34fb4f3dc4e1d55ed4140

File: lib/mint/http1.ex (hunk 2 above)
Comment (gdwoolbert3) [reply lib/mint/http1.ex]: Ahh, interesting! I didn't know about the pin being preferred there.

My initial assumption was that, if the initial decode didn't fail, it wouldn't fail with the sentinel bit. That said, your approach is much more explicit.
5940b9a68d479de238e34fb4f3dc4e1d55ed4140

File: test/mint/http1/conn_test.exs (hunk 4 above)
Comment (gdwoolbert3) [reply test/mint/http1/conn_test.exs]: 5940b9a68d479de238e34fb4f3dc4e1d55ed4140

File: lib/mint/http1.ex (hunk 5 above)
Comment (gdwoolbert3) [reply lib/mint/http1.ex]: 5940b9a68d479de238e34fb4f3dc4e1d55ed4140

File: lib/mint/http1.ex (hunk 2 above)
Comment (gdwoolbert3) [reply lib/mint/http1.ex]: I made one small tweak here in regards to the usage of `with` and `case` but the control flow is the same:
9c223c686ddb61802f9e2f16fbd00b7f74353d2d

**Review by ericmj — APPROVED** [review]

File: lib/mint/http1.ex [hunk 6]
```diff
@@ -744,7 +751,7 @@ defmodule Mint.HTTP1 do
   end
 
   defp decode_headers(conn, request, data, responses, headers) do
-    case Response.decode_header(data) do
+    case do_decode_header(data, conn.stream_headers) do
```
Comment (whatyouhide) [inline lib/mint/http1.ex]: **Nit**: the `do_` prefix has disappeared from Elixir conventions years ago. Can we rename this to `decode_header/2` or `decode_single_header/2` if that's taken or something like that?

File: lib/mint/http1.ex [hunk 7]
```diff
@@ -758,7 +765,12 @@ defmodule Mint.HTTP1 do
       {:ok, :eof, rest} ->
         case add_header_bytes(conn, request, byte_size(data) - byte_size(rest)) do
           {:ok, request} ->
-            responses = [{:headers, request.ref, Enum.reverse(headers)} | responses]
+            responses =
+              case {conn.stream_headers, headers} do
+                {true, []} -> responses
+                {_, headers} -> [{:headers, request.ref, Enum.reverse(headers)} | responses]
+              end
```
Comment (whatyouhide) [inline lib/mint/http1.ex]: I think this reads significantly better as:


```suggestion
              if conn.stream_headers and headers == [] do
                responses
              else
                [{:headers, request.ref, Enum.reverse(headers)} | responses]
              end
```

File: lib/mint/http1.ex (hunk 6 above)
Comment (gdwoolbert3) [reply lib/mint/http1.ex]: c2036c34cf10912bd1776d0f1f0868ab2000fc81

File: lib/mint/http1.ex (hunk 7 above)
Comment (gdwoolbert3) [reply lib/mint/http1.ex]: c2036c34cf10912bd1776d0f1f0868ab2000fc81

**Comment by gdwoolbert3** [comment]
@whatyouhide apologies, I had to update the line reference in the dialyzer ignore file. I suspect that the workflow should pass now

**Comment by whatyouhide** [comment]
Thank you @gdwoolbert3! 🫶
